docs/okeps/okep-5212-ovnobserv-api.md

```diff
@@ -0,0 +1,241 @@
+to "subscribe" to the samples enabled by this configuration.
+
+### Workflow Description
+
+1. Cluster admin wants to enable OVN Observability for admin-managed features.
+2. Cluster admin created a CR with 100% probabilities for (Baseline)AdminNetworkPolicy and EgressFirewall.
+3. Messages like "Allowed by admin network policy test, direction Egress" appear on netobserv UI.
+4. Cluster admin sees CPU usage by ovnkube pods grows up 10%, and reduces the sampling probability to 50%.
+5. Still most of the flows have correct enrichment with lower overhead.
+6. Admin is happy. Until...
+7. A user from project "bigbank" reports that they can't access the internet from pod A.
+8. Cluster admin enables debugging for project "bigbank" with 100% probability for NetworkPolicies.
+9. netobserv UI or ovnkube-oberv script shows a sample with <`source pod A`, `dst_IP`> saying "Denied by network policy in namespace bigbank, direction Egress"
+10. Admin now has to convince the user that they are missing an allow egress rule to `dst_IP` for pod A.
+11. Network policy is fixed, debugging can be turned off.
+
+### API Details
+
+#### Observability filtering for debugging
+
+Mostly for debugging (but also can be used for observability if needed), we may need more granular filtering,
+like per-node or per-namespace filtering.
+
+- Per-node filtering may be implemented by adding a node selector to `ObservabilityConfig`, and is easy to implement in ovn-k
+  as every node has its own ovn-k instance with OVN-IC.
+- Per-namespace filtering can only be applied to the namespaced Observability Features, like NetworkPolicy and EgressFirewall.
+  Cluster-scoped features, like (Baseline)AdminNetworkPolicy, can only be additionally filtered based on name. Label selectors
+  can't be used, as observability module doesn't watch any objects.
+
+#### ObservabilityConfig CRD
+
+The CRD for this feature will be a part of ovn-kubernetes to allow using it upstream.
+
+```go
+
+// ObservabilityConfig described OVN Observability configuration.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=observabilityconfig,scope=cluster
+// +kubebuilder:singular=observabilityconfig
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+type ObservabilityConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ObservabilitySpec `json:"spec"`
+	// +optional
+	Status ObservabilityStatus `json:"status,omitempty"`
+}
+
+type ObservabilitySpec struct {
+	// CollectorID is the ID of the collector that should be unique across the cluster, and will be used to receive configured samples.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=1
+	// +required
+	CollectorID int32 `json:"collectorID"`
```

Comment:
> So when this CR is created, couldn't OVNK assign an ID, rather than the user having to provide one?

we could do that, but it will require a centralized (aka cluster-manager) watcher to assign IDs. Then we will need to report that ID back via status, so that it can be used with debug script.
I am not sure if this is worth an extra centralized component to automate this part, wdyt?

Reply:
Could you not make `ObservabilityStatus` something like this?

```
// ObservabilityStatus contains the observed status of the ObservabilityConfig.
type ObservabilityStatus struct {
        // A collection of CollectorIDs, indexed by nodeName
        CollectorIDs map[string]int32 
	// +listType=map
	// +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}
```

Then it's easy enough for users to look up the ID by node name.